File: 괄호회전하기.py
def solution(s):
    answer = 0
    
    for i in range(len(s)):
        palindrome = s[i: ] + s[0: i]
        
        stk = []
        stopped = False
        for pal in palindrome:
            if pal == '(' or pal == '[' or pal =='{':
                stk.append(pal)
            else:
                # 닫힌 괄호가 매칭이 안되는 경우 || 괄호가 없는 경우
                if len(stk) == 0:
                    stopped = True
                    break
                
                p = stk.pop()
                
                if not (p == '(' and pal == ')' or p == '{' and pal == '}' or p == '[' and pal == ']'):
                    stopped = True
                    break
                
                
        if not stopped and len(stk) == 0: # 모든 괄호가 매칭된 경우
            answer += 1 
            
        
        
    return answer

File: 괄호회전하기.java
import java.util.*;

class Solution {
    public int solution(String s) {
        int answer = 0;
        
        int N = s.length();
        for(int i=0; i<N; i++){
            String palindrome = s.substring(i, N) + s.substring(0, i);
            
            boolean stopped = false;
            List<Character> stk = new ArrayList<>();
            for(char pal: palindrome.toCharArray()){ // string to char[] 
                if(pal == '(' || pal == '[' || pal =='{')
                    stk.add(pal);
                else{
                    if(stk.size() == 0){
                        stopped = true;
                        break;
                    }
                    char p = stk.remove(stk.size()-1);
                    
                    if(! (p == '(' && pal == ')' || 
                          p == '{' && pal == '}' || 
                          p == '[' && pal == ']')){
                        stopped = true;
                        break;
                    }
                }
            }
            
            
            if(!stopped && stk.isEmpty())
                answer++;
            
            
            
        }
        
        
        
        return answer;
    }
}
